src/library.js

```diff
@@ -3027,9 +3035,119 @@ LibraryManager.library = {
   msync: function(addr, len, flags) {
     // int msync(void *addr, size_t len, int flags);
     // http://pubs.opengroup.org/onlinepubs/009696799/functions/msync.html
-    // Pretend to succeed
+    // TODO: support sync'ing parts of allocations
+    var info = _mmap.mappings[addr];
+    if (!info) return 0;
+    if (len == info.num) {
+      var buffer = new Uint8Array(HEAPU8.buffer, addr, len);
+      return FS.msync(FS.getStream(info.fd), info, buffer, 0, len, flags);
+    }
+
     return 0;
   },
+  
+  // ==========================================================================
+  // emscripten_mmap.h
+  // ==========================================================================
+  mmap_read: function(heap, addr, addrOnHeap, type, index) {
+    var info = _mmap.mappings[addr];
+    // ?? How to handle this? 0 is a weird return value in a way...
+    if (!info) return 0;
+    return FS.mmap_read(FS.getStream(info.fd), info, type, index, heap[addrOnHeap + index]);
+  },
+  mmap_write: function(heap, addr, addrOnHeap, type, index, value) {
+    var info = _mmap.mappings[addr];
+    if (!info) return -1;
+    // Always write to heap too
+    heap[addrOnHeap + index] = value;
```

Comment:
Writing to the heap is harmless here. Prevents MEMFS from handling this, and if you're talking against node then it doesn't matter as you'll never see this value.